src/vs/workbench/contrib/terminalContrib/chatAgentTools/browser/runInTerminalTool.ts

```diff
@@ -418,6 +463,89 @@ export class RunInTerminalTool extends Disposable implements IToolImpl {
 		}
 	}
 
+	private async _pollForOutputAndIdle(
+		execution: BackgroundTerminalExecution,
+		extendedPolling: boolean,
+		token: CancellationToken
+	): Promise<{ idle: boolean; output: string; pollDurationMs?: number }> {
+		const maxWaitMs = extendedPolling ? 120000 : 20000;
+		const maxInterval = 2000;
+		let currentInterval = 500;
+		const pollStartTime = Date.now();
+
+		let lastBufferLength = 0;
+		let idleCount = 0;
+		let buffer = '';
+
+		while (true) {
+			const now = Date.now();
+			const elapsed = now - pollStartTime;
+			const timeLeft = maxWaitMs - elapsed;
+
+			if (timeLeft <= 0) {
+				break;
+			}
+
+			// Cap the wait so we never overshoot timeLeft
+			const waitTime = Math.min(currentInterval, timeLeft);
+			await timeout(waitTime);
+
+			// Check again immediately after waking
+			if (Date.now() - pollStartTime >= maxWaitMs) {
+				break;
+			}
+
+			currentInterval = Math.min(currentInterval * 2, maxInterval);
+
+			buffer = execution.getOutput();
+			const currentBufferLength = buffer.length;
+
+			if (currentBufferLength === lastBufferLength) {
+				idleCount++;
+			} else {
+				idleCount = 0;
+				lastBufferLength = currentBufferLength;
+			}
+			const isLikelyFinished = await this._assessOutputForFinishedState(buffer, token);
+			if (isLikelyFinished && idleCount >= 2) {
+				return { idle: true, output: buffer, pollDurationMs: Date.now() - pollStartTime + (extendedPolling ? 20000 : 0) };
+			}
```

Comment:
On "idle" not being clear again, I was thinking it mean it's idling as in no data events for x seconds and not finished.